### PR TITLE
Test Case and fixes for DNSHInfo

### PR DESCRIPTION
**Pull request by stephenrauch**

### CODE REVIEW

**Comment by jstasiak** [comment]
Thank you @stephenrauch. There seem to be multiple commits repeated in this pull request, #48 and #45, many of which don't seem relevant to the pull requests they're part of.

May I ask you to leave only the relevant changes in the pull requests, squash commits that belong together (some of the commits break tests and are followed up by commits that fix the issues) and break other commits (that contain multiple unrelated changes, like https://github.com/jstasiak/python-zeroconf/pull/49/commits/75232ccf28a820ee723db072951078eba31145a5) into a smaller ones?

I'd also very much appreciate providing a more detailed description of what's being done in each commit (for example it's not clear to me what's being fixed in https://github.com/jstasiak/python-zeroconf/pull/49/commits/79d0c1eea8f75971361fa072fe767f6eb0ca123d).

**Comment by jstasiak** [comment]
Oh, I forgot: inheriting ``unittest.TestCase`` is not necessary (the project uses nose to run tests) and in general putting tests in a class is unnecessary too. Additionally it's quite important to use custom assertion functions in order to get meaningful error messages (nose doesn't provide nice assertion messages out of the box like py.test does), so for example ``eq_(value, expected)`` rather than ``assert value == expected`` (``eq_`` comes from http://nose.readthedocs.org/en/latest/testing_tools.html).

**Comment by stephenrauch** [comment]
The repeated changes in this PR are there only because I built them on top of the branch from the previous PR as you noted.  That code had yet to be merged, and I really needed to build on top of the previous code because building on top of the code prior to that branch was causing enough errors to be annoying.  If the previous pull requests are merged, then this PR only has two commits present on this pull request. Those two submissions were kept separate because one was just formatting changes. The progression is pretty easy to see here:

https://github.com/jstasiak/python-zeroconf/network

I tried very hard to keep the check-ins monolithic. I have already spent considerable time squashing commits to discrete chucks.  I do not believe there are any commits that broke existing test cases.  If you are referring to 75232cc, I believe that failure was actually latent, and the new test cases I added, simply exposed the failure, which was a race condition, and had nothing to do with my test case. So I did not squash the check in with the test case in question because it is an independent bug.

As a practical matter, 3 1/2 weeks after I started this work there really is no way to modify this progression at this point, short of squashing it all together, and that would be even more confusing as you noted in your commentary on 75232cc.

As to inheriting from ```unittest.TestCase```, more of the code in the test cases uses this model than not.  So that is what I adopted.  I would suggest that if you don't want the code written this way, then the entirety of the unit tests could be converted.

I don't mind submitting an additional PR to change the asserts to a more specific format needed by nose.  Let me know if you would like me to do that. Have you considered using pytest instead?  As you noted, it has less need for these special assert methods to be informative.

Finally, for the three PRs in question, I wrote about 800 words of change description, so I am really unsure how I could be more detailed. This PR is a bit thin, but I frankly thought the changes in this PR to be documented in the title. The DNSHInfo packet generation was just broken.  There was no test case for that functionality, and adding a test case showed four issues. Two of which were relative to unicode, one of which was a reference to an attribute which just didn't exist, and finally the fields used the wrong encoding which is what necessitated the change from write_string() to write_character_string().  So, looking back at it, the encoding failure is not self evident from the changes, but the other three are pretty straight forward in doing a review of the change itself.

If you have any specific questions I would be thrilled to answer them, but I don't feel that a blanket statement of "add more detail" is a fair assessment of the work I did here.

Cheers,

**Comment by jstasiak** [comment]
Out of all those commits first few I had looked at included:
* https://github.com/jstasiak/python-zeroconf/pull/49/commits/79d0c1eea8f75971361fa072fe767f6eb0ca123d
* https://github.com/jstasiak/python-zeroconf/pull/49/commits/d909942e2c9479819e9113ffb3a354b1d99d6814

Neither of those is self-explanatory to me and that's what I was referring to regarding additional description. The rest of the commits is documented very well, my apologies if I implied otherwise.

File: zeroconf.py [hunk 1]
```diff
@@ -1080,6 +1080,7 @@ def __init__(self, type, name, address=None, port=None, weight=0,
             self.server = server
         else:
             self.server = name
+        self._properties = {}
```
Comment (stephenrauch) [inline zeroconf.py]: Race condition in which field is used before it is declared.  Latent bug exposed bug new test case.

**Comment by stephenrauch** [comment]
@jstasiak There several Pull Requests pending. Do you have a plan to get them merged anytime soon? Can I be of help?

**Pull request merged**